Get acls subcommand

petedannemann opened the pull request:
This PR adds the `get acls` subcommand. It's behavior was based heavily on Kafka's built in kafka-acls.sh --list script.
It supports filtering on various fields so that users can answer questions related to which services have access to which
resources.

This PR also fixes flaky test failures [like this one ](https://github.com/segmentio/topicctl/actions/runs/6474554914/job/17579712456) by adding some sleep time in a test


Examples:
```bash
$ topicctl get acls --cluster-config examples/local-cluster/cluster.yaml --help
Displays information for ACLs in the cluster. Supports filtering with flags.

Usage:
  topicctl get acls [flags]

Examples:
List all acls
$ topicctl get acls

List read acls for topic my-topic
$ topicctl get acls --resource-type topic --resource-name my-topic --operations read

List acls for user Alice with permission allow
$ topicctl get acls --principal User:alice --permission-type allow

List acls for host 198.51.100.0
$ topicctl get acls --host 198.51.100.0


Flags:
  -h, --help                                help for acls
      --host string                         The host to filter on. (e.g. 198.51.100.0)
      --operation ACLOperationType          The operation that is being allowed or denied to filter on. allowed: "any", "all", "read", "write", "create", "delete", "alter", "describe", "clusteraction", "describeconfigs", "alterconfigs" or "idempotentwrite" (default any)
      --permission-type ACLPermissionType   The permission type to filter on. allowed: "any", "allow", or "deny" (default any)
      --principal string                    The principal to filter on in principalType:name format (e.g. User:alice).
      --resource-name string                The resource name to filter on. (e.g. my-topic)
      --resource-pattern-type PatternType   The type of the resource pattern or filter. allowed: "any", "match", "literal", "prefixed". "any" will match any pattern type (literal or prefixed), but will match the resource name exactly, where as "match" will perform pattern matching to list all acls that affect the supplied resource(s). (default any)
      --resource-type ResourceType          The type of resource to filter on. allowed: "any", "topic", "group", "cluster", "transactionalid", "delegationtoken" (default any)
```
```bash
$ topicctl get acls --cluster-config examples/local-cluster/cluster.yaml
[2023-09-15 18:28:46]  INFO ACLs:
----------------+--------------+---------------+------------+---------+-----------+------------------
  RESOURCE TYPE | PATTERN TYPE | RESOURCE NAME | PRINCIPAL  |  HOST   | OPERATION | PERMISSION TYPE
----------------+--------------+---------------+------------+---------+-----------+------------------
  group         | literal      | junk3         | User:foo5  | 1.2.3.4 | all       | deny
  group         | literal      | junk3         | User:foo5  | 1.2.3.7 | all       | deny
  group         | literal      | junk3         | User:foo2  | 1.2.3.4 | read      | allow
  topic         | prefixed     | team1_        | User:team2 | *       | all       | deny
  topic         | literal      | junk          | User:foo5  | 1.2.3.4 | all       | deny
  topic         | literal      | junk          | User:foo2  | 1.2.3.4 | read      | allow
  topic         | literal      | junk          | User:foo2  | 1.2.3.4 | describe  | allow
  topic         | literal      | junk          | User:foo5  | 1.2.3.7 | all       | deny
  topic         | literal      | foo           | User:alice | *       | describe  | allow
  topic         | literal      | foo           | User:alice | *       | create    | allow
  topic         | literal      | foo           | User:alice | *       | write     | allow
----------------+--------------+---------------+------------+---------+-----------+------------------
$ topicctl get acls --cluster-config examples/local-cluster/cluster.yaml --resource-type topic --resource-pattern-type literal --resource-name junk --permission-type allow --operation describe
[2023-09-15 18:28:30]  INFO ACLs:
----------------+--------------+---------------+-----------+---------+-----------+------------------
  RESOURCE TYPE | PATTERN TYPE | RESOURCE NAME | PRINCIPAL |  HOST   | OPERATION | PERMISSION TYPE
----------------+--------------+---------------+-----------+---------+-----------+------------------
  topic         | literal      | junk          | User:foo2 | 1.2.3.4 | describe  | allow
----------------+--------------+---------------+-----------+---------+-----------+------------------
```

petedannemann commented on go.mod:
```diff
@@ -11,7 +11,7 @@ require (
 	github.com/olekukonko/tablewriter v0.0.5
 	github.com/samuel/go-zookeeper v0.0.0-20190923202752-2cc03de413da
-	github.com/segmentio/kafka-go v0.4.35
+	github.com/segmentio/kafka-go v0.4.43-0.20230913165112-9ecb9d2f7da5
```
Need to release a new version of kafka-go with the added ACLs APIs and bump the version here

petedannemann commented on pkg/admin/brokerclient.go:
```diff
@@ -690,3 +696,69 @@ func configEntriesToAPIConfigs(
+}
+
+// CreateACL creates an ACL in the cluster.
+func (c *BrokerAdminClient) CreateACL(
```
This can't be used by the user yet but this will be required in future work and adding it now made testing much better

petedannemann commented on pkg/admin/types.go:
```diff
@@ -73,6 +75,221 @@ type PartitionAssignment struct {
+// We need to subtype this to be able to define methods to
+// satisfy the Value interface from Cobra so we can use it
+// as a Cobra flag.
+type ResourceType kafka.ResourceType
```
These types allow us to convert between kafka-go types (most of which are int8s that represent something in the Kafka API) and human read able forms of the types for the user to provide through the CLI or see displayed as output

hhahn-tw reviewed (approved):
Looks good. A cosmetic suggestion for your help strings that list allowed values: quoting each makes it harder to read in my opinion, how about something like this instead?
```
The permission type to filter on. allowed: [any, allow, deny]
```

petedannemann commented:
@hhahn-tw since your last review I added repl support, bumped to the released version of kafka-go and changed `CreateACL` to `CreateACLs` as that will be more useful in later PRs

hhahn-tw commented on pkg/cli/repl.go:
```diff
@@ -39,6 +39,10 @@ var (
 	getSuggestions = []prompt.Suggest{
+		{
+			Text:        "acls",
+			Description: "Get all acls",
```
"acls" should be capitalized here and below at line 471 for consistency with other text output

petedannemann replied:
Good catch, fixed

hhahn-tw reviewed (changes requested):
Text style change request "acl" -> "ACL" in info output to user
